node/src/components/chain_synchronizer/operations.rs

```diff
@@ -948,34 +954,147 @@ where
+        + From<FetcherRequest<BlockAndDeploys>>
+        + From<StorageRequest>
+        + From<ContractRuntimeRequest>,
 {
     let start = Timestamp::now();
-    let fetched_block = fetch_retry_forever::<_, BlockAndDeploys>(ctx, block_hash).await?;
-    let res = match fetched_block {
-        FetchedData::FromStorage {
-            item: block_and_deploys,
-            ..
-        } => Ok(block_and_deploys),
-        FetchedData::FromPeer {
-            item: block_and_deploys,
-            ..
-        } => {
-            ctx.effect_builder
-                .put_block_and_deploys_to_storage(block_and_deploys.clone())
-                .await;
-            Ok(block_and_deploys)
-        }
-    };
-    ctx.metrics
-        .observe_fetch_block_and_deploys_duration_seconds(start);
-    res
+    'a: loop {
+        let fetched_block = fetch_retry_forever::<_, BlockAndDeploys>(ctx, block_hash)
+            .await
+            .map_err(FetchAndStoreBlockError::Fetcher)?;
+        let res = match fetched_block {
+            FetchedData::FromStorage {
+                item: block_and_deploys,
+                ..
+            } => Ok(block_and_deploys),
+            FetchedData::FromPeer {
+                item: block_and_deploys,
+                ..
+            } => {
+                if let Some(deploy_approvals_root_hash) =
+                    fetch_deploy_approvals_root_hash(block_and_deploys.block.height(), ctx).await?
+                {
+                    let approvals: BTreeMap<DeployHash, BTreeSet<Approval>> = block_and_deploys
+                        .deploys
+                        .iter()
+                        .map(|d| (*d.id(), d.approvals().clone()))
+                        .collect();
+                    let mut approval_hashes = vec![];
+                    for (dh, maybe_approvals) in block_and_deploys
+                        .block
+                        .body()
+                        .deploy_hashes()
+                        .iter()
+                        .chain(block_and_deploys.block.body().transfer_hashes().iter())
+                        .map(|dh| (*dh, approvals.get(dh)))
+                    {
+                        if let Some(approvals) = maybe_approvals {
+                            approval_hashes.push(Digest::hash(
+                                approvals
+                                    .to_bytes()
+                                    .map_err(FetchAndStoreBlockError::BytesRepr)?,
+                            ));
+                        } else {
+                            error!("no approvals for deploy {:?}", dh);
+                            return Err(FetchAndStoreBlockError::NoApprovalsForDeploy(dh));
+                        }
+                    }
+
+                    if !approval_hashes.is_empty()
+                        && Digest::hash_merkle_tree(approval_hashes) != deploy_approvals_root_hash
+                    {
+                        continue 'a;
```

Comment:
Should we disconnect from the peer?

Reply:
I'm not sure, is that what we do in other places if something is off?

Reply:
Yep, it seems appropriate to disconnect here based on the rest of the code.

Reply:
Yes, in general if a peer is sending us clearly wrong data (that _must_ be their fault), we disconnect.